torch/distributed/__init__.py

```diff
@@ -54,6 +56,27 @@ def is_available() -> bool:
         _make_nccl_premul_sum,
     )
 
+    def breakpoint(rank: int = 0):
+        """
+        Set a pdb breakpoint, but only on a single rank.  All other ranks will wait for you to be
+        done with the pdb session before continuing.
+        """
+        if get_rank() == rank:
+            # This will be the case when your subprocess was created by
+            # multiprocessing.Process, see
+            # https://stackoverflow.com/questions/30134297/python-multiprocessing-stdin-input
+            old_stdin = None
+            if isinstance(sys.stdin, io.TextIOWrapper):
+                old_stdin = sys.stdin
+                sys.stdin = open(0)
+            try:
+                pdb.set_trace(header=f"\n!!! ATTENTION !!!\n\nType 'up' to get to the frame that called dist.breakpoint(rank={rank})\n")
+            finally:
+                if old_stdin is not None:
+                    sys.stdin.close()
+                    sys.stdin = old_stdin
+        barrier()
```

Comment:
i've thought a bit about whether to use a barrier or not in this pattern.  it is a bit awkward either way.

case 1- you want to step forward on the debugger and you encounter a collective. it will hang as it waits for other ranks which are stuck on the barrier
case 2- you don't use the barrier- other ranks will proceed beyond where you break and probably hang on some collective (or exit).  

I think its better to have the barrier.  Maybe also a print saying that if you want to continue, to also call barrier() on the debugger before continuing?

Reply:
or did i misunderstand- does 'continuing' from this actually let the breakpointed rank call the barrier()?  IIRC i tried that at one point and it didn't work, so i didn't PR it.  but maybe i did something subtly wrong.

Reply:
But you don't have to call barrier on the debugger. If you single step past the breakpoint() spot, you will run the barrier logic. This seems fine to me? The big problem is if you advance forward, now all the other processes will race to the next collective, and now they could potentially timeout waiting for you to finish your session. But afaict, there isn't really any good way around this.

Reply:
> does 'continuing' from this actually let the breakpointed rank call the barrier()?

I believe that's how it works yeah.

Reply:
ok. yea, i am not able to think of a better solution either, just maybe a warning/info print that lets people understand what's going on.

Reply:
I can't think of a good way to do this warning. I would need to somehow see if I was crossing the barrier from inside a pdb session, this doesn't seem like available API surface.